RemoteTransform and RemoteTransform2D enhancements

TwistedTwigleg opened the pull request:
Added the following features to RemoteTransform and RemoteTransform2D:
- Added the ability to toggle changes in position, rotation, and scale
- Added the ability to toggle changes in either local or global coordinates

closes #265

TwistedTwigleg committed:
Added the following features to RemoteTransform and RemoteTransform2D:
- Added the ability to toggle changes in position, rotation, and scale
- Added the ability to toggle changes in either local or global coordinates

akien-mga commented:
Note that there seems to be something weird in your local git configuration, your commits are shown as "TwistedTwigleg committed with = 7 hours ago". So it's like you author the commits with a git identity linked to your GitHub account (TwistedTwigleg), and then reapply those commits (git am/rebase) with a different git identity named "=".

akien-mga commented on scene/2d/remote_transform_2d.cpp:
```diff
@@ -63,7 +63,21 @@ void RemoteTransform2D::_update_remote() {
 
 	//todo make faster
-	n->set_global_transform(get_global_transform());
+	if (use_full_transform == true) {
```
You don't need to check for `== true`, `if (use_full_transform)` works fine already.

TwistedTwigleg commented:
Okay, I made the changes based on the review.

I'm not 100% sure why all of my commits are with a git identity named "=". It seems "=" was set as the default local git user, though I used --author in the most recent commit, which should override local.

I just used 'git --unset-all user.name' and 'git --unset-all user.email' and now it seems the git identity named "=" has been removed.

Should I see if I can alter the previous commits to remove references of the git user named "="? Or should I just leave it as is?

akien-mga commented:
If you rebase your branch according to http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase to squash the two commits together (second one as fixup of the first one), it should automatically update the committer reference too.

You can check your ~/.gitconfig file (or the corresponding one on Windows/macOS, not sure where it would be) to ensure that "=" is no longer in it.

TwistedTwigleg commented:
Okay, will do! Just changed branches and am quickly fixing an issue there, then I switch back to this branch and will squish the two commits together.

I'm still trying to wrap my head around git, so I appreciate all of the help! :smiley: Thanks!

nunodonato commented:
Did you test all cases? I get some weird behavior

TwistedTwigleg commented:
I tested all of the new options, and they seemed to work.

I'll admit I didn't have a project to use for testing, so it's possible I missed something, or messed something up.

What behavior are you seeing?

nunodonato commented:
"use full transform" means it applies the changes in global coordinates instead of local, right? If so, I dont think "full transform" is clear on what it does. I'd suggest "use global coordinates" or something similar.

I need to test this better, but on toggle things on/off sometimes I got in cases where the rotation would happen even if rotation was off.

TwistedTwigleg commented:
Yeah, using full transform means it applies in global coordinates. I'll switch 'use full transform' to 'use global coordinates', I think it sounds better and should reduce confusion.

The rotation issue might be because I'm updating the remote node on option toggling. I'll need to do some more testing to be sure, but I think that could be the problem.

TwistedTwigleg commented:
Alright, I think I found and fixed the rotation issue. I renamed "use_full_transform" to "use_global_coordinates", thanks nunodonato for the suggestion!

I also squished the commits into one commit.

nunodonato commented:
yep, works ok now! thanks :)

TwistedTwigleg commented:
No problem, I'm glad it works.
Thanks for the suggestion and the bug report!

bojidar-bg commented on scene/2d/remote_transform_2d.cpp:
```diff
@@ -63,7 +63,21 @@ void RemoteTransform2D::_update_remote() {
+		if (update_remote_rotation)
+			n->set_global_rotation(get_global_rotation());
+		if (update_remote_scale)
+			n->set_global_scale(get_global_scale() + Vector2(1, 1));
```
I feel that you should do `n->set_global_transform(get_global_transform());` if all of the three options are checked.
That way it would work for shears and other strange transforms (which can't be represented as position + rotation + scale).

bojidar-bg replied:
Additionally, I think you shouldn't be adding `(1, 1)` to the scale...

bojidar-bg replied:
Additionally, if only rotation and scale should be copied, you can still pull this off, with something like:
```cpp
if (update_remote_rotation && update_remote_scale) {
	Transfrom2D global_transform = get_global_transform();
	if (!update_remote_position)
		global_transform.origin = Vector2();
	n->set_global_transform(global_transform);
} else {
	… // Current code goes here
}
```

TwistedTwigleg replied:
I'll make changes so that set_global_transform is used if all options are true. I forgot about transforms like shears, whoops!

On using global transform for rotation and scale, won't switching the origin to (0, 0) cause issues? Then the remote node's position will be set to (0, 0), right? I could get the node's global transform and change origin to the node's origin. That might be the way to do it. I'll do some testing and see what happens!

TwistedTwigleg replied:
I added (1, 1) to scale because I found that nodes had their scale set to zero (or rather 0.00001)  if the remote transform had a scale of (1, 1) and adding one fixed it. Probably a bug in my code.

bojidar-bg replied:
Ah, right, my bad, you are indeed supposed to get the current origin 😄

bojidar-bg commented on scene/2d/remote_transform_2d.cpp:
```diff
@@ -63,7 +63,21 @@ void RemoteTransform2D::_update_remote() {
+		if (update_remote_rotation)
+			n->set_rotation(get_global_rotation());
+		if (update_remote_scale)
+			n->set_scale(get_global_scale());
```
Don't copy global position/scale/rotation, copy local ones :wink:

TwistedTwigleg replied:
Whoops! 😂

bojidar-bg commented on scene/2d/remote_transform_2d.cpp:
```diff
@@ -102,6 +116,41 @@ NodePath RemoteTransform2D::get_remote_node() const {
 
+void RemoteTransform2D::set_use_global_coordinates(const bool p_enable) {
+	use_global_coordinates = p_enable;
+}
```
Forgot `_update_remote()` call?

TwistedTwigleg replied:
This was intentional.

I wasn't sure if I should update the remote based on changes in coordinates, since this can lead to the remote's transform being changed by accident (say someone accidentlly toggles the wrong option) and since it edits the node's transform, it could be a hassle to move back to it's original position. I don't think Ctrl-Z could undo the transform change. I'll test, and if it does I'll make the change for sure

I'm not against changing it though, but I thought I'd explain the reasoning behind why I left it out. I'll change it if you think it's not going to be a problem or undo works.

bojidar-bg replied:
Right, makes sense.

bojidar-bg commented on scene/3d/remote_transform.cpp:
```diff
@@ -102,6 +121,41 @@ NodePath RemoteTransform::get_remote_node() const {
 
+void RemoteTransform::set_use_global_coordinates(const bool p_enable) {
+	use_global_coordinates = p_enable;
+}
```
`_update_remote();`

TwistedTwigleg replied:
See comment above

bojidar-bg commented on scene/3d/remote_transform.cpp:
```diff
@@ -116,11 +170,32 @@ void RemoteTransform::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::NODE_PATH, "remote_path"), "set_remote_node", "get_remote_node");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "use_global_coordinates"), "set_use_global_coordinates", "get_use_global_coordinates");
+
+	ADD_GROUP("Update", "");
```
I think you might pass "change_" as the second parameter. (or rename the properties to start with "update_", and pass that)

TwistedTwigleg replied:
Okay, sounds good. I think I'll rename the properties to start with "_update".

bojidar-bg reviewed (changes requested):
Looks pretty good, you might want to do some of those:

TwistedTwigleg commented:
Okay, I changed the code:
- Uses transforms when possible, instead of setting position, rotation, and scale.
- Changed the group from `ADD_GROUP("Update", "")` to `ADD_GROUP("Update, "update_")`
- If all of the options are selected, then it uses the exact transform

I left in this code in remote_transform_2d:
```
if (update_remote_scale)
    n->set_global_scale(get_global_scale() + Vector2(1, 1));
```
Because without it when using global coordinates the scale of the remote turns to (0.0001, 0.00001) despite the fact that Remote2D's scale is (1, 1). Not sure if it is a bug in my code, or if it's a bug in node_2D set_scale function.

TwistedTwigleg commented:
I changed the method of updating the remote so if there is any special transforms they always get passed to the remote node.

bojidar-bg commented on scene/2d/remote_transform_2d.cpp:
```diff
@@ -63,7 +63,44 @@ void RemoteTransform2D::_update_remote() {
+			n->set_global_transform(our_trans);
+
+			if (update_remote_scale)
+				n->set_global_scale(get_global_scale() + Vector2(1, 1));
```
Why isn't this handled as the two above?

TwistedTwigleg replied:
Setting transform sets scale too, and I couldn't find any way to change a transform's scale. So I set the scale after to get around it.

TwistedTwigleg commented:
@bojidar-bg is everything good here?

bojidar-bg commented:
Well, yes, my points have been handled (sorry for not changing review status earlier). Would be nice if someone else also goes over the code, anyway. ( @tagcup Would you be able to do this, since you tend to understand matrices and transforms better? 😃 )

akien-mga commented:
A rebase could also be interesting, I don't think we need 4 different commits for the overall diff: http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase

TwistedTwigleg commented:
No worries bojidar-bg, I just wanted to make sure everything in your review was handled.

akien-mga I'll rebase and squish the commits into one, sorry about that.

TwistedTwigleg commented:
I fixed a couple bugs I missed with scaling on RemoteTransform2D.

While I was fixing these bugs, I looked at Node2D's set_global_scale function and I'm a little confused.
Set_global_scale seems to take the parent node's global scale and subtract that from the scale you pass in. Why does it subtracts the parent's scale?

(Side note: It would appear that the parent's scale was being subtracted from the scale I passed in, hence why in the previous commit I had to add (1, 1) to the scale for it to stay synced. Mystery solved :sunglasses: )

Another thing I found is that set_global_scale is never used in the code base (according to a Atom project find using "set_global_scale"). If it's never used, shouldn't we remove it? It's just dead code and I'm not sure it even works correctly. I don't think you can set global scale without changing the parent node(s) since global scale is effected by the parent node(s).

(I'll squish the commits together after someone gives it a look. I'm not 100% sure everything is working correctly)

bojidar-bg commented:
@TwistedTwigleg It is bound to GDScript, even if not used by C++.

bojidar-bg commented:
Also, note that set_global_scale comes from #6159, and was modeled after set_global_rot, it seems.

If you can make it using Transform2D-s, it would be all for the better, I think.

TwistedTwigleg commented:
@bojidar-bg RemoteTransform2D uses Tranform2D-s, and RemoteTransform uses Transform-s

I think I've figured out what's happening with set_global_scale.

Here's the equation for figuring out a node's global scale
```
global_scale = our_scale * parent_node_scale;
```
Here is what I think set_global_scale is currently doing:
```
our_scale = global_scale - parent_node_scale;
```
Which doesn't get the global scale. This method does work for getting global rotation though, because global rotation is:
```
global_rotation = our_rotation + parent_node_rotation;
```
And setting global rotation is:
```
our_rotation = global_rotation - parent_node_rotation
```


To get global scale though, you'd need this:
```
our_scale = global_scale / parent_node_scale;
```
Right?

Should I open a pull request fixing this, assuming my logic is correct?

bojidar-bg commented:
@TwistedTwigleg Sounds good, so, yeah, you should probably make a PR fixing this.

TwistedTwigleg commented:
@bojidar-bg
I made the pull request: #9486

TwistedTwigleg commented:
I squished the five commits into one
